### ISSUE TITLE
Sequence-Wise Batch Normalization for RNN's

### ISSUE THREAD

**Issue by LeavesBreathe** [tensorflow/tensorflow#917]
Hey TF,

Recently, for deep RNN's, sequence wise batch normalization has proven to be very helpful. [Deep Speech 2](http://arxiv.org/pdf/1512.02595v1.pdf) in section 3.2 explains this in more detail. Sequence-wise batch normalization is described in section 4.1 in [Batch Normalized RNNs](http://arxiv.org/pdf/1510.01378v1.pdf).

`tf.nn.moments` is very useful for batch normalization because it gives you the mean and variance. However, in seq2seq setups, we need to do a batch normalization across all timesteps (the entire sequence). Unfortunately, the way `seq2seq.py` is written, each timestep is within a separate 2D matrix. This means that there is a list of 2d tensors for the entire sequence.

If somehow, `tf.nn.moments` could be modified to accept lists of tensors (all of the same dimensions), it would be incredibly helpful. This way, we could input the entire list of tensors, and compute the resulting mean and variance for that sequence. 

Thanks!

**Comment by ludimagister** [comment]
This is a good idea, we put this on the list of things to do.

**Comment by LeavesBreathe** [comment]
Thanks really appreciate it!

**Comment by vincentvanhoucke** [comment]
@LeavesBreathe I'm looking into some of these issues right now. Is there any specific reason why a `tf.nn.moments([] ...)` type of API would be preferable to simply doing `tf.nn.moments(tf.pack([]) ...)` ?

**Comment by LeavesBreathe** [comment]
If I understand things correctly, tf.pack would copy all the tensors, and require more memory. Please correct me if I'm wrong. 

Otherwise, your right: this is a much better solution. Thanks for the help.

**Comment by vincentvanhoucke** [comment]
@LeavesBreathe in the grand scheme of things, it will not affect memory consumption much: as soon as the copy is performed, the input buffers will no longer be needed for the backward pass, and be releasable to the pool. Packing things into one tensor also potentially enables more efficient contractions to take place. What matters most to memory consumption is long-lived tensors, e.g. ones that have to be preserved between the forward and the backward pass. One could still consider a different API if there is evidence that this is a problem.
I'm also looking into separating the collection of sufficient statistics (which can be performed online) from the actual aggregation, which would give users the freedom to make the reductions more granular if they wanted. I'll be closing this bug now, and treat the other items as opportunities for optimization.

**Comment by LeavesBreathe** [comment]
Thank you vincent, I will try this approach and apologize for not thinking of it sooner!

**Comment by vincentvanhoucke** [comment]
I have just committed a change to `nn.moments()` which allows you to decouple the collection of the sufficient statistics `nn.sufficient_statistics()` from the computation of the means and variances `nn.normalize_moments()`. This allows you to build compact statistics at each time step, aggregate them by summing them whichever way you want, and normalize them afterwards. I hope this helps.

**Comment by LeavesBreathe** [comment]
@vincentvanhoucke I have been away for a while but I'm trying to implement batch norm now on LSTMs. 

Thank you for updating `nn.moments()` -- I will try to work on this to implement sequence wise batch normalization. Really appreciate the help!
